example/particles/particles.cpp

```diff
@@ -309,46 +424,179 @@ TaskStatus Defrag(MeshBlock *pmb) {
 
   // Only do this if list is getting too sparse. This criterion (whether there
   // are *any* gaps in the list) is very aggressive
-  if (s->get_num_active() <= s->get_max_active_index()) {
+  if (s->GetNumActive() <= s->GetMaxActiveIndex()) {
     s->Defrag();
   }
 
   return TaskStatus::complete;
 }
 
-// See the advection_driver.hpp declaration for a description of how this function gets
-// called.
-TaskCollection ParticleDriver::MakeTaskCollection(BlockList_t &blocks, int stage) {
+// Custom step function to allow for looping over MPI-related tasks until complete
+TaskListStatus ParticleDriver::Step() {
+  TaskListStatus status;
+  integrator.dt = tm.dt;
+
+  BlockList_t &blocks = pmesh->block_list;
+  auto num_task_lists_executed_independently = blocks.size();
+
+  // Create all the particles that will be created during the step
+  status = MakeParticlesCreationTaskCollection().Execute();
+
+  // Loop over repeated MPI calls until every particle is finished. This logic is
+  // required because long-distance particle pushes can lead to a large, unpredictable
+  // number of MPI sends and receives.
+  bool particles_update_done = false;
+  while (!particles_update_done) {
+    status = MakeParticlesUpdateTaskCollection().Execute();
+
+    particles_update_done = true;
+    for (auto &block : blocks) {
+      // TODO(BRR) Despite this "my particles"-specific call, this function feels like it
+      // should be generalized
+      auto swarm = block->swarm_data.Get()->Get("my particles");
+      if (!swarm->finished_transport) {
+        particles_update_done = false;
+      }
+    }
+  }
+
+  // Use a more traditional task list for predictable post-MPI evaluations.
+  status = MakeFinalizationTaskCollection().Execute();
+
+  return status;
+}
+
+// TODO(BRR) This should really be in parthenon/src... but it can't just live in Swarm
+// because of the loop over blocks
+TaskStatus StopCommunicationMesh(const BlockList_t &blocks) {
+  int num_sent_local = 0;
+  for (auto &block : blocks) {
+    auto &pmb = block;
+    auto sc = pmb->swarm_data.Get();
+    auto swarm = sc->Get("my particles");
+    swarm->finished_transport = false;
+    num_sent_local += swarm->num_particles_sent_;
+  }
+
+  // Boundary transfers on same MPI proc are blocking
```

Comment:
I think this comment only makes sense if the `TODO(BRR) May want logic like this if we have non-blocking TaskRegions` code underneath is uncommented. There, we check that boundary communications are complete, but we don't have to do that for same-MPI-proc communications because those can be assumed to be already done by the time we're here. Since this comment is only for theoretical code, I've removed it.